lib/cdo/aws/cloudfront.rb

```diff
@@ -0,0 +1,211 @@
+        log: {
+          bucket: 'cdo-logs',
+          prefix: "#{ENV['RACK_ENV']}-pegasus-cdn"
+        }
+      },
+      dashboard: {
+        aliases: [CDO.dashboard_hostname],
+        origin: "#{ENV['RACK_ENV']}-dashboard.code.org",
+        ssl_cert: 'codeorg-cloudfront',
+        log: {
+          bucket: 'cdo-logs',
+          prefix: "#{ENV['RACK_ENV']}-dashboard-cdn"
+        }
+      }
+    }
+
+    # Creates or updates the CloudFront distribution based on the current configuration.
+    # Calls to this method should be idempotent, however CloudFront distribution updates can take ~15 minutes to finish.
+    #
+    # Ref: http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/HowToUpdateDistribution.html
+    # "Your changes don't propagate to every edge location instantaneously; propagation to all edge locations can take
+    #  15 minutes. When propagation is complete, the status of your distribution changes from InProgress to Deployed.
+    #  While CloudFront is propagating your changes to edge locations, we cannot determine whether a given edge
+    #  location is serving your content based on the previous configuration or the new configuration."
+    def self.create_or_update
+      cloudfront = Aws::CloudFront::Client.new
+      ids = %i(pegasus dashboard).map do |app|
+        distribution = cloudfront.list_distributions.distribution_list.items.detect do |i|
+          i.aliases.items.include?(CDO.method("#{app}_hostname").call)
+        end
+        id = nil
+        if distribution
+          id = distribution.id
+          distribution_config = cloudfront.get_distribution_config(id: id)
+          cloudfront.update_distribution(
+            id: id,
+            if_match: distribution_config.etag,
+            distribution_config: config(app, distribution_config.distribution_config.caller_reference)
+          )
+          puts "#{app} distribution updated!"
+        else
+          resp = cloudfront.create_distribution(
+            distribution_config: config(app)
+          )
+          id = resp.distribution.id
+          puts "#{app} distribution created!"
+        end
+        [app, id]
+      end
+
+      ids.map do |app, id|
+        cloudfront.wait_until(:distribution_deployed, id: id) do |waiter|
+          waiter.before_wait { |_| puts "Waiting for #{app} distribution to deploy.." }
+        end
+        puts "#{app} distribution deployed!"
+      end
+    end
+
+    # Returns a CloudFront DistributionConfig Hash compatible with the AWS SDK for Ruby v2.
+    # Syntax reference: http://docs.aws.amazon.com/sdkforruby/api/Aws/CloudFront/Types/DistributionConfig.html
+    # `app` is a symbol containing the app name (:pegasus or :dashboard)
+    def self.config(app, reference = nil)
+      config = HTTP_CACHE[app]
+      cloudfront = CONFIG[app]
+      behaviors = config[:behaviors].map do |behavior|
+        if behavior[:path].is_a? Array
+          behavior[:path].map do |path|
```

Comment:
Could you refactor to avoid duplicating the code for the array and non-array case?  (e.g. wrap in an array if not already an array)